### PR TITLE
[#74] Overhaul package storage

### LINKED ISSUE #74: Path should depend on the package name
For example: `data/hackage/packet/x`

### PR DESCRIPTION
**Pull request by neongreen**
Fixes #74.

### CODE REVIEW

File: core/src/main/scala/codesearch/core/index/SourceDownloader.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import sys.process._
+
+import scala.concurrent.Future
+
+case class TargetPackage(
+    name: String,
+    version: String
+)
+
+trait SourceDownloader {
```
Comment (Som3) [inline core/src/main/scala/codesearch/core/index/SourceDownloader.scala]: mb SourceRepository? in future, it will enable create multiple repository implementations - e.g. Java - mvn and jfrog
name SourceRepository more clearly imho

Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/SourceDownloader.scala]: This will be changed in the following commits.

File: Makefile [hunk 2]
```diff
@@ -1,6 +1,7 @@
 .RECIPEPREFIX +=
 
 # Run Postgres
+.PHONY: db
```
Comment (neongreen) [inline Makefile]: Rebase please

File: core/src/main/scala/codesearch/core/index/NpmIndex.scala [hunk 3]
```diff
@@ -12,7 +12,7 @@ import play.api.libs.json.Json
 
 import scala.collection.mutable
 
-object NpmIndex extends Index with NpmDB {
+object  NpmIndex extends Index with NpmDB {
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/NpmIndex.scala]: Accidentally?

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+case class NpmPackage(
+    rawName: String,
+    version: String
+) extends SourcePackage {
+  //Because package name can look like: react>>=native@@router!!v2.1.123(refactored:-))
+  val name: String                     = URLEncoder.encode(rawName, "UTF-8")
+  val fsPath: Path                     = Paths.get(new URI(s"./npm/$name/$version"))
+  val url: URI                         = URI.create(s"https://registry.npmjs.org/$name/-/$name-$version.tgz")
+  override def extentions: Set[String] = Set("js", "json", "xml", "yml", "coffee", "markdown", "md", "yaml", "txt")
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: typo: `extensions`
(maybe it's not your typo, but still)

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      Seq("tar", "-xvf", archive.getCanonicalPath, "-C", directory.getCanonicalPath) !!;
+      directory
+    }
+}
+
+trait SourcePackage extends Extractor {
+  def fsPath: Path
+  def url: URI
+  def extentions: Set[String] = Set.empty
+}
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: Comments please (if you don't know what these fields mean then ask around)

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+) extends SourcePackage {
+  //Because package name can look like: react>>=native@@router!!v2.1.123(refactored:-))
+  val name: String                     = URLEncoder.encode(rawName, "UTF-8")
+  val fsPath: Path                     = Paths.get(new URI(s"./npm/$name/$version"))
+  val url: URI                         = URI.create(s"https://registry.npmjs.org/$name/-/$name-$version.tgz")
+  override def extentions: Set[String] = Set("js", "json", "xml", "yml", "coffee", "markdown", "md", "yaml", "txt")
+}
+
+trait Download[A <: SourcePackage] {
+  def downloadSources(pack: A): Future[Int]
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: `pack` = `package`?

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  val url: URI     = URI.create(s"https://hackage.haskell.org/package/$name-$version/$name-$version.tar.gz")
+}
+
+case class GemPackage(
+    name: String,
+    version: String
+) extends SourcePackage {
+  val fsPath: Path                     = Paths.get(new URI(s"./gem/$name/$version"))
+  val url: URI                         = URI.create(s"https://rubygems.org/downloads/$name-$version.gem")
+  override def extentions: Set[String] = Helper.langByExt.keySet
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: Why is `langByExt` used here and but not in other places?

(NB: I don't know whether it's better to get rid of `langByExt` and inline it everywhere, or do the opposite)

File: core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      to
+    }
+    ???
+  }
+
+  def unzip(file: File): Future[File] = {
+    ???
+  }
+
+  def deleteExcessFiles(directory: File, allowedExtentions: Set[String]): Future[Int] = Future {
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala]: Comments please

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 4 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: Thnx, i fix it.

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 5 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: It is value by default for trait. Implementations override this field.

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 7 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: I did by analogy with existing functions

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 6 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: Yes. In Scala word `package` already used by language.

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 7 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: I will check this logic

File: core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala (hunk 8 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala]: The function filters files leaving only those whose extensions contains in allowable set.

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 5 above)
Comment (neongreen) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: I meant comments like "`extentions` specifies which files from the package will be retained after the package is downloaded; this is useful to filter out binaries and other artifacts that occasionally occur in package archives"

btw, why is it a part of the trait at all? shouldn't it be per-language, not per-package?

File: core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala (hunk 8 above)
Comment (neongreen) [reply core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala]: Comments in code, not on Github :slightly_smiling_face:

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 6 above)
Comment (Som3) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: can be used `_`
e.g. `type` -> `_type` | `type_`

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala (hunk 7 above)
Comment (Som3) [reply core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: @neongreen we can create enums with all meta information about language - e.g. extensions

I wanted to do it in https://github.com/aelve/codesearch/issues/71

File: core/src/main/scala/codesearch/core/index/Sources.scala [hunk 9]
```diff
@@ -39,80 +37,39 @@ trait Sources[VTable <: DefaultTable] {
...
-      })
-    }.map(_.sum)
+    val futureAction = indexAPI
+      .verNames()
+      .flatMap { packages =>
+        val packagesMap = Map(packages: _*)
+
+        Future.sequence(lastVersions.filter {
+          case (packageName, currentVersion) =>
+            !packagesMap.get(packageName).contains(currentVersion)
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/Sources.scala]: It is slightly more "scalish" to use `filterNot` without `!` inside.

Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/Sources.scala]: Thnx, it is all will be refactored.

File: core/src/main/scala/codesearch/core/index/CratesSources.scala [hunk 10]
```diff
@@ -1,58 +1,41 @@
...
       val nameToVersion = Map(verSeq: _*)
-      (answers.length, answers
-        .slice(math.max(page - 1, 0) * 100, page * 100)
-        .flatMap(uri => indexAPI contentByURI(uri, nameToVersion)).groupBy(x => (x._1, x._2)).map {
-        case ((name, packageLink), results) =>
-          PackageResult(name, packageLink, results.map(_._3))
-      }.toSeq.sortBy(_.name))
+      (answers.length,
+       answers
+         .slice(math.max(page - 1, 0) * 100, page * 100)
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/CratesSources.scala]: May be better declare PAGE_SIZE in `Sources` scope or whatever in order to avoid logical or mistype errors in the future.

**Comment by neongreen** [comment]
@kamilongus sorry for lots of conflicts :(

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    unzippingMethod(archive, unarchived)
+    moveAllFromDir(unarchived)
+  }
+
+  /**
+    *
+    * @param unarchived is directory contains unarchived files
+    * @return same directory containing all files and directories from unarchived files
+    */
+  def moveAllFromDir(unarchived: File): File = {
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Should files in `unarchived` folder be moved too? Current algorithm doesn't move them.

Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Method name is incomprehensible. May be better something like "flatDir".

File: core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    *
+    * @param directory is root directory of specific package
+    * @param allowedExtentions is extensions defined for each language
+    * @return count removed files
+    */
+  private def deleteExcessFiles(directory: File, allowedExtentions: Set[String]): Future[Int] = Future {
+    def filterFiles(all: List[File], excess: List[File] = Nil): List[File] = all match {
+      case Nil => excess
+      case file :: others if file.isDirectory =>
+        if (file.listFiles().length > 0)
```
Comment (dem4nd) [inline core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala]: 1. It is better to save file.listFiles in val. 
2. listFiles has no side effect and brackets are redundant.
3. `file.listFiles().length > 0` has concise form `file.listFiles.nonEmpty`

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala (hunk 11 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: I agree

File: core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala (hunk 12 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/SourceRepository.scala]: 1. Ok;
2. IMHO, in theory I/O operations have side effects, but in this case we can omit it;
3. Ok.

File: core/src/main/scala/codesearch/core/db/DefaultDB.scala [hunk 13]
```diff
@@ -2,20 +2,26 @@ package codesearch.core.db
...
 
+import codesearch.core.index.repository.SourcePackage
 import codesearch.core.model._
-import slick.lifted.TableQuery
 import slick.jdbc.PostgresProfile.api._
+import slick.lifted.TableQuery
 
-import scala.concurrent.Future
 import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/db/DefaultDB.scala]: Can we make scalafmt sort imports automatically?

File: core/src/main/scala/codesearch/core/index/LanguageIndex.scala [hunk 14]
```diff
@@ -141,41 +99,34 @@ trait LanguageIndex[VTable <: DefaultTable] { self: DefaultDB[VTable] =>
...
+    *
     * @return last versions of packages
     */
   protected def getLastVersions: Map[String, Version]
 
   /**
-    * download source code from remote repository
+    * Update source code from remote repository
+    *
+    * @see [[https://github.com/aelve/codesearch/wiki/%D0%A1odesearch-developer-Wiki#updating-packages]]
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/LanguageIndex.scala]: I fixed the link, please use `Codesearch-developer-Wiki#updating-packages`

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+private[index] trait Extractor {
+
+  /**
+    * @param from is file to unarchiving
+    * @param to is target directory
+    */
+  def unzippingMethod(from: File, to: File): Unit =
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: If `to` is a directory, why does it say `File`? I assume it should be `Path`

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    ArchiverFactory
+      .createArchiver(TAR, GZIP)
+      .extract(from, to)
+
+  /**
+    * @param archive is file to unarchiving
+    * @param directory is target directory
+    * @return directory containing all unarchived files and directories
+    */
+  def extract(archive: File, directory: Path): Future[File] = Future {
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: and the return type should also be `Path`, I guess, or something

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+  /**
+    * @param unarchived is directory contains unarchived files
+    * @return same directory containing all files and directories from unarchived files
+    */
+  def flatDir(unarchived: File): File = {
+    unarchived
+      .listFiles()
+      .filter(_.isDirectory)
+      .foreach(_.listFiles().foreach(file =>
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Does `listFiles()` look into hidden files by default?

File: core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala [hunk 18]
```diff
@@ -0,0 +1,49 @@
+package codesearch.core.index.directory
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala]: This module is missing docstrings

File: core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+object PackageDirectory {
+
+  implicit class PackageDirectoryOps[A <: SourcePackage](val pack: A) {
+    def archive(implicit ev: Directory[A]): Path    = ev.archive(pack)
+    def unarchived(implicit ev: Directory[A]): Path = ev.unarchived(pack)
+  }
+
+  implicit def hackageDirectory: Directory[HackagePackage] = new Directory[HackagePackage] {
+    override def archive(pack: HackagePackage): Path =
+      root / "hackage" / s"${pack.name}" / s"${pack.version}" / s"${pack.name}-${pack.version}.$extension"
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala]: Why does it say `root` here? I guess we want relative paths, or I'm not sure what we want – but definitely not putting everything into the server root

File: core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  implicit def npmDirectory: Directory[NpmPackage] = new Directory[NpmPackage] {
+    override def archive(pack: NpmPackage): Path =
+      root / "npm" / s"${pack.name}" / s"${pack.version}" / s"${pack.name}-${pack.version}.$extension"
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala]: Will this work with crazy NPM package names?

I don't know how `NpmPackage` is defined but it would make sense to store "true" names there and do the escaping here or in some other place (as long as we don't have five different places where the logic for finding a package is reimplemented).

Maybe we can do that in a separate PR.

File: core/src/main/scala/codesearch/core/index/repository/Extensions.scala [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+
+trait Extension[A] {
+  def extensions: Set[String]
+}
+
+object Extensions {
+
+  implicit def jsExtensions[A <: JavaScript]: Extension[A] = new Extension[A] {
+    override def extensions: Set[String] =
+      Set("js", "ts", "json", "xml", "yml", "coffee", "markdown", "md", "yaml", "txt")
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/Extensions.scala]: Hm, this is actually wrong. We need two sets of extensions – one to figure out which files to leave after downloading the package, and another set to determine which files are counted as "source files" as far as the relevant checkbox is concerned.

If my concern is valid, please create a separate issue for that.

File: core/src/main/scala/codesearch/core/index/repository/Extensions.scala [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+  implicit def haskellExtensions[A <: Haskell]: Extension[A] = new Extension[A] {
+    override def extensions: Set[String] = Set("hs", "lhs", "hsc", "hs-boot", "lhs-boot")
+  }
+
+  implicit def rubyExtensions[A <: Ruby]: Extension[A] = new Extension[A] {
+    override def extensions: Set[String] = Set("rb", "rbx", "irb")
+  }
+
+  implicit def rustExtensions[A <: Rust]: Extension[A] = new Extension[A] {
+    override def extensions: Set[String] = Set("rs", "cabal")
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/Extensions.scala]: Can you remove `cabal` from here?

File: core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+) extends SourcePackage with Rust {
+  val url: Uri = uri"https://crates.io/api/v1/crates/$name/$version/download"
+}
+
+final case class NpmPackage(
+    rawName: String,
+    version: String
+) extends SourcePackage with JavaScript {
+  //Because package name can look like: react>>=native@@router!!v2.1.123(refactored:-))
+  val name: String = URLEncoder.encode(rawName, "UTF-8")
```
Comment (neongreen) [inline core/src/main/scala/codesearch/core/index/repository/SourcePackage.scala]: Oh yes, apparently we don't store raw names. This could be handled better but let's tackle this in another PR.

File: core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala (hunk 19 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala]: `root` value defined in trait Directory by default `./data`

Comment (neongreen) [reply core/src/main/scala/codesearch/core/index/directory/PackageDirectory.scala]: alright, but then let's rename it to e.g. `dataRoot` or something, because otherwise it can be mixed up with `Ammonite.root`

File: core/src/main/scala/codesearch/core/db/DefaultDB.scala (hunk 13 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/db/DefaultDB.scala]: We already have it. We use `asciisortimports`.
More patterns here: https://scalameta.org/scalafmt/#SortImports

File: core/src/main/scala/codesearch/core/index/repository/Extensions.scala (hunk 22 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/repository/Extensions.scala]: Ok

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala (hunk 15 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Ok

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala (hunk 16 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Ok

File: core/src/main/scala/codesearch/core/index/directory/Extractor.scala (hunk 17 above)
Comment (kamilongus) [reply core/src/main/scala/codesearch/core/index/directory/Extractor.scala]: Yes it does.

File: core/src/main/scala/codesearch/core/index/repository/Extensions.scala (hunk 21 above)
Comment (neongreen) [reply core/src/main/scala/codesearch/core/index/repository/Extensions.scala]: > please create a separate issue for that

I've done that (#93)
